### ISSUE TITLE
Verify matching of border modes for DL4J vs Keras

### ISSUE THREAD

**Issue by maxpumperla** [deeplearning4j/deeplearning4j#3935]
#### Issue Description

We currently match both "valid" and "full" border modes  to `ConvolutionMode.Truncate`. Verify that this is correct.

Valid mode:
TensorFlow and Theano "valid" modes apply filter only to complete patches within the image borders with no padding. That is equivalent to DL4J Truncate mode with no padding.

Full mode:
Theano-only "full" mode zero pads the image so that outputs = (inputs + filters + 1) / stride. This should be equivalent to DL4J Truncate mode with padding equal to filters-1.

**Comment by maxpumperla** [comment]
@AlexDBlack what's your take on this? I'm not convinced our mapping is correct. Are you familiar with keras border modes?

**Comment by AlexDBlack** [comment]
Right, Valid==Truncate, Same==Same.

I think full should be (in+k-1)/s. Truncate does (in-k+2*p)/s+1, which does not batch the Full equation. If we set p=k-1, then I  think we can map to truncate.
http://deeplearning.net/software/theano/tutorial/conv_arithmetic.html#full-padding
